Add Software Development Contract - internal (#44)
bornfight/README: playbooks/software-development-internal-contract.md

---
title: Software Development Contract - internal
description:
    This document is designed to explain in what manner will software dev. dept. members contribute to a project.
---

## Functional specification

> -   An official document without which the development shouldn’t be started

Reviewer: ```suggestion
-   An official document as an absolute requirement for development to start
```

-   It provides an answer to:

>     -   Which software functionalities are we going to develop?

Reviewer: ```suggestion
    -   Which software functionalities are to be developed?
```

-   By using it:
    -   Differences between client expectations, and our implementation can be greatly decreased
    -   The chances of misunderstanding between people involved in a project can be decreased

>     -   There is a “single source of truth” to which all parties can rely upon if a misunderstanding occurs

Reviewer: ```suggestion
    -   There is a “single source of truth” for reference in case of misunderstanding
```

>     -   A clear difference can be seen between “functionalities agreed upon" and "change requests"

Reviewer: ```suggestion
    -   A change request is easier to define
```

    -   Efficiency increases because we can predict potential problems from the very start of software development

>     -   A need for back-and-forth communication between a client and us is decreased

Reviewer: ```suggestion
    -   A need for back-and-forth communication between the client and our side is decreased
```

    (Jira, Teamwork...)
-   These tasks should be permanently available and transparent
-   One software functionality should represent one _feature task_
-   Feature task should be broken down into feature subtasks
    -   They should be created by project Tech Lead per department involved in the development
    -   They have to be approved and confirmed by PO
-   Feature tasks and subtasks should be described following BDD principles, so all participants can understand
    them
    -   example:
        [How Behavior Driven Development increases project quality](https://www.bornfight.com/blog/behavior-driven-development-how-to-find-bugs-before-coding-starts/)
-   Feature subtasks can be broken down into even smaller tasks called technical or implementation tasks
-   All task-relevant discussion and communication should be done in task comments
-   Every task and subtask should be described using scenarios which can be used as acceptance criteria

### How to break down a functionality

#### Example software feature “Clients”

-   Functionality: Insert a new client

    -   Actors: client, user, tablet application
    -   Requirements:

>         -   I, as a user, have a new client, and I want to store his information. This functionality is used in two

Reviewer: ```suggestion
        -   I, as a user, have a new client, and I want to store this client's information. This functionality is used in two
```

        represent clients interest in the best manner (functional specification should be considered as a client’s
        best interest)
    -   Other responsibilities can be found in
        [How Behavior Driven Development increases project quality](https://www.bornfight.com/blog/behavior-driven-development-how-to-find-bugs-before-coding-starts/)
        article
    -   What can you expect from the person performing this role:
        -   To take responsibility for the end project delivery
        -   To be responsible for functional specification creation and for a functional specification to feature
            task break-down
    -   What you can’t expect from the person performing this role:
        -   To take responsibility for software delivery
        -   To solve software related problems

### Software Development project team

-   Technical Lead
    -   Is familiar with the project requirements and software implementation and infrastructure
    -   Exclusively responsible for software delivery
    -   Senior-level or Middle-level developers
    -   Is in charge of task break down

>     -   _This role brings with it an increased level of responsibility for the software delivery in a predicted

Reviewer: ```suggestion
    -   _This role comes with increased level of responsibility for the software delivery in a predicted
```

        deadline_
    -   What can you expect from the person performing this role:
        -   To take responsibility for software delivery
        -   To be familiar with software business logic in detail

>         -   To know the answer to all of your questions about software requirements, business logic or technical

Reviewer: ```suggestion
        -   To know answers to all of your questions about software requirements, business logic or technical
```

            implementation details
        -   To assist in functional specification creation
        -   To do mayor estimates
        -   To break down feature tasks into the subtasks
        -   To conceptualize software architecture and in need, implement it
        -   To support makers on the project (provide answers about business logic, help with implementation...)
        -   To solve software related problems
        -   To be a code owner of the software until that ownership is transferred or code is obsoleted
    -   What you can’t expect from the person performing this role:
        -   To take responsibility for the end project delivery
        -   To be in charge of direct client communication
        -   To take care of the project budget
        -   To be completely responsible for functional specification creation and for a software feature to task
            break-down
        -   To guide project makers through every step of the feature implementation
-   Maker
    -   All seniority levels
    -   A person that is in charge of implementing features based on feature tasks and subtasks

>     -   Has a lot less responsibility for software delivery

Reviewer: ```suggestion
    -   Delivery responsibility is mainly focused on tasks and features, depending on the maker seniority level
```

    -   What can you expect from the person performing this role:
        -   To do feature implementation based on the tasks and subtasks
        -   To estimate feature implementation

>         -   To communicate its road-block, impediments or requests help

Reviewer: ```suggestion
        -   To communicate roadblocks, impediments and ask for help
```

        -   To have a responsibility towards the Tech. Lead for feature implementation
    -   What you can’t expect from the person performing this role:

>         -   Complete independence while doing its work

Reviewer: ```suggestion
        -   Complete autonomy while working
```

>         -   To take responsibility for software delivery

Reviewer: ```suggestion
        -   To take the final responsibility for software delivery
```
